server/remote_cache/cachetools/cachetools.go

```diff
@@ -86,8 +93,39 @@ func GetBlob(ctx context.Context, bsClient bspb.ByteStreamClient, r *digest.Reso
 		wc = decompressor
 	}
 
+	type streamMsg struct {
+		rsp *bspb.ReadResponse
+		err error
+	}
+	streamMsgs := make(chan streamMsg)
+	go func() {
+		for {
+			rsp, err := stream.Recv()
+			select {
+			case streamMsgs <- streamMsg{rsp, err}:
+			case <-ctx.Done():
+				return
+			}
+			if err != nil {
+				return
+			}
+		}
+	}()
+	readTimer := time.NewTimer(*casRPCTimeout)
+	defer readTimer.Stop()
+
 	for {
-		rsp, err := stream.Recv()
+		var rsp *bspb.ReadResponse
+		var err error
+		select {
+		case msg := <-streamMsgs:
+			rsp = msg.rsp
+			err = msg.err
+		case <-readTimer.C:
+			return status.DeadlineExceededError("Timed out waiting for Read response")
+		case <-ctx.Done():
+			return ctx.Err()
```

Comment:
sorta dumb question -- could we use context.WithTimeout(ctx, *flagVal) if the context timeout is unset or longer, and then remove the readTimer?

Reply:
Had to rework some things, but I was able to work the context timeout into this.